pkg/client/selfhosted/selfhosted.go

```diff
@@ -0,0 +1,232 @@
+package selfhosted
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/jetstack/version-checker/pkg/api"
+)
+
+const (
+	// This n=500 is a temporary work around until pagination is properly tested
+	// Not all versions support pagination AND not all Artifactory versions are handling the "latest" argument
+	lookupURL   = "%s/%s/%s/tags/list?n=500"
+	manifestURL = "%s/%s/%s/manifests/"
+)
+
+type Options struct {
+	URL      string
+	LoginURL string
+	Username string
+	Password string
+	Bearer   string
+}
+
+type Client struct {
+	*http.Client
+	Options
+}
+
+type AuthResponse struct {
+	Token string `json:"token"`
+}
+
+type TagResponse struct {
+	Name string   `json:"name"`
+	Tags []string `json:"tags"`
+}
+
+type Result struct {
+	Name      string  `json:"name"`
+	Timestamp string  `json:"last_updated"`
+	Images    []Image `json:"images"`
+}
+
+type Image struct {
+	Digest       string `json:"digest"`
+	OS           string `json:"os"`
+	Architecture string `json:"Architecture"`
+}
+
+type ManifestResponse struct {
+	SchemaVersion int        `json:"schemaVersion"`
+	MediaType     string     `json:"mediaType"`
+	Manifests     []Manifest `json:"manifests"`
+}
+
+type Manifest struct {
+	MediaType string   `json:"mediaType"`
+	Size      int      `json:"size"`
+	Digest    string   `json:"digest"`
+	Platform  Platform `json:"platform"`
+}
+
+type Platform struct {
+	Architecture string `json:"architecture"`
+	Os           string `json:"os"`
+}
+
+func New(ctx context.Context, opts Options) (*Client, error) {
+	client := &http.Client{
+		Timeout: time.Second * 5,
+	}
+
+	// Only try to setup auth if an actually URL is present.
+	if opts.URL != "" {
+		// Setup Auth if username and password used.
+		if len(opts.Username) > 0 || len(opts.Password) > 0 {
+			if len(opts.Bearer) > 0 {
+				return nil, errors.New("cannot specify Bearer as well as username/password")
+			}
+
+			token, err := basicAuthSetup(client, opts)
+			if err != nil {
+				return nil, fmt.Errorf("failed to setup auth: %s", err)
+			}
+			opts.Bearer = token
+		}
+	}
+
+	return &Client{
+		Options: opts,
+		Client:  client,
+	}, nil
+}
+
+func (c *Client) Tags(ctx context.Context, _, repo, image string) ([]api.ImageTag, error) {
+	url := fmt.Sprintf(lookupURL, c.Options.URL, repo, image)
+	urlManifest := fmt.Sprintf(manifestURL, c.Options.URL, repo, image)
+	var tags []api.ImageTag
+
+	response, err := c.doRequest(ctx, url)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, tag := range response.Tags {
+
+		manifestResponse, err := c.doManifestRequest(ctx, urlManifest+tag)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, manifest := range manifestResponse.Manifests {
+
+			sha := strings.Replace(manifest.Digest, "sha256:", "", 1)
+
+			tags = append(tags, api.ImageTag{
+				Tag:          tag,
+				SHA:          sha,
+				Timestamp:    time.Now(),
```

Comment:
This can be done in another PR but the timestamp is discoverable, but as a backwards compatible object when not setting the header  `application/vnd.docker.distribution.manifest.list.v2+json`

```
         "v1Compatibility": "{\"architecture\":\"amd64\",\"config\":{\"Hostname\":\"\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"Tty\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":[\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"],\"Cmd\":null,\"Image\":\"sha256:be438750e9b044019d71cc9ac4241d48917bad2484574451a8239f4071587f30\",\"Volumes\":null,\"WorkingDir\":\"\",\"Entrypoint\":[\"/usr/bin/version-checker\"],\"OnBuild\":null,\"Labels\":{\"description\":\"Kubernetes utility for exposing used image versions compared to the latest version, as metrics.\"}},\"container\":\"7b5179890fa494425bed03a1baf096d9fe2cde7d2e2d58e4af51e1f155ecab4a\",\"container_config\":{\"Hostname\":\"7b5179890fa4\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"Tty\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":[\"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin\"],\"Cmd\":[\"/bin/sh\",\"-c\",\"#(nop) \",\"ENTRYPOINT [\\\"/usr/bin/version-checker\\\"]\"],\"Image\":\"sha256:be438750e9b044019d71cc9ac4241d48917bad2484574451a8239f4071587f30\",\"Volumes\":null,\"WorkingDir\":\"\",\"Entrypoint\":[\"/usr/bin/version-checker\"],\"OnBuild\":null,\"Labels\":{\"description\":\"Kubernetes utility for exposing used image versions compared to the latest version, as metrics.\"}},\"created\":\"2020-08-22T16:35:32.362045622Z\",\"docker_version\":\"19.03.12-ce\",\"id\":\"49948de4091ff7fb42b24da8a053fba0a0e46d47a5ce495b5e7f11bd66bd7fbd\",\"os\":\"linux\",\"parent\":\"c4c337344d46dd4b25346ce6072246541e6ed9fcc6f504c6d3f6b889cf9c2916\",\"throwaway\":true}"
```

I'm still not decided on whether we wan't to make _another_ call per digest, just for the timestamp. Perhaps we can only do it if we are searching for digests.

Reply:
With my latest commit, relates to comment: https://github.com/jetstack/version-checker/pull/35#discussion_r475932955 this is now also fixed.